packages/hw-transport-node-hid/src/listenDevices.js

```diff
@@ -27,37 +37,93 @@ export default (
 
   let lastDevices = getFlatDevices();
 
-  const checkDevices = () => {
+  const poll = (type, attempt = 1) => {
+    let changeFound = false;
+
     if (!listenDevicesPollingSkip()) {
+      debug(`Polling for ${type} [attempt ${attempt}/${MAX_ATTEMPTS}]`);
+
       const currentDevices = getFlatDevices();
 
-      const newDevices = currentDevices.filter(d => !lastDevices.includes(d));
-      const removeDevices = lastDevices.filter(
-        d => !currentDevices.includes(d)
-      );
+      if (type === "add") {
+        const newDevices = currentDevices.filter(d => !lastDevices.includes(d));
+
+        if (newDevices.length > 0) {
+          debug("New device found:", newDevices);
+
+          listDevices = getDevices();
+          events.emit("add", getDeviceByPaths(newDevices));
 
-      if (newDevices.length > 0) {
-        listDevices = getDevices();
-        events.emit("add", getDeviceByPaths(newDevices));
+          changeFound = true;
+        } else {
+          debug("No new device found");
+        }
       }
 
-      if (removeDevices.length > 0) {
-        events.emit("remove", getDeviceByPaths(removeDevices));
-        listDevices = listDevices.filter(
-          d => !removeDevices.includes(flatDevice(d))
+      if (type === "remove") {
+        const removeDevices = lastDevices.filter(
+          d => !currentDevices.includes(d)
         );
+
+        if (removeDevices.length > 0) {
+          debug("Removed device found:", removeDevices);
+
+          events.emit("remove", getDeviceByPaths(removeDevices));
+          listDevices = listDevices.filter(
+            d => !removeDevices.includes(flatDevice(d))
+          );
+
+          changeFound = true;
+        } else {
+          debug("No removed device found");
+        }
+      }
+
+      if (changeFound) {
+        lastDevices = currentDevices;
+      } else {
+        if (attempt < MAX_ATTEMPTS) {
+          const newDelay = delay * attempt;
+
+          debug(`Repolling ${type} in ${newDelay}ms`);
+
+          setTimeout(() => {
+            poll(type, attempt + 1);
+          }, newDelay);
+        } else {
+          debug(`Giving up after ${attempt} attempts`);
+        }
       }
+    } else {
+      debug(`Polling skipped, retrying in ${delay}ms`);
 
-      lastDevices = currentDevices;
+      setTimeout(() => {
+        poll(type);
```

Comment:
we have to poll multiple times because event happen too soon? Is it always the case, if so it would be great to estimate a minimal time and do an initial timeout to likely never have to poll twice. 🤔  

I think it would maybe preferable to refactor things using lodash `debounce` . each time there are remove/add events you just "push to 50ms later" the re-poll() , and the else case also is about triggering a poll().

e.g.

`poll = debounce(() => { }, delay)`

not sure what delay will be , if it make sense to be the same 1s that was before 🤔

Reply:
> we have to poll multiple times because event happen too soon?

From my tests I have to poll a second time (after 500ms) only for device removal on macOS and Windows, but I have no idea how much this behavior is consistent from machine to machine 🤔